### ISSUE TITLE
[Site] multiples observers/digitisers pour un site

### ISSUE THREAD

**Issue by MathRdt** [PnX-SI/gn_module_monitoring#233]
Bonjour,

# Objectif

Dans le cadres de développements pour le CREA avec @ColinVanReeth , nous souhaiterions partager des sites entre plusieurs personnes afin que les-dits sites restent "privés" (au sens permissions CRUVED, "je ne peux voir que mes sites" par exemple).

Les personnes concernées peuvent faire partie d'organismes différents et donc la solution "je ne peux voir que les sites de mon organisme" ne fonctionne pas dans notre cas d'usage.

# Solution proposée

Une solution qui me semblerait ne pas prendre trop d'énergie, serait de copier ce qui a été fait au niveau d'une visite, et de créer une table de corrélation entre observers et sites pour que plusieurs personnes puissent partager des droits sur un même site, sans qu'ils appartiennent au même organisme ou aient des permissions spécifiques qui leurs permettraient de voir l'entièreté des sites du sous-module (ce qui n'est pas notre objectif ici).

Qu'en pensez vous ?

**Comment by camillemonchicourt** [comment]
Hum, le système de permissions actuel avec le CRUVED et les portées me semble déjà suffisamment complexe à gérer, maintenir et comprendre, pour encore ajouter un autre volet et fonctionnement qui s'y ajoute.
Et il faudrait pouvoir gérer le contenu de cette table, en plus du système complexe de permissions ?

**Comment by MathRdt** [comment]
Je ne pense pas que cela rajoute grand chose en termes de complexité, comme le travail est déjà fait pour les visites.

Il faut juste rajouter une table `cor_site_observer`, update le modèle de la table `t_base_sites` pour avoir la liste d'observers, ainsi que les quelques lignes côté API/Front pour remplir cette table lors de la saisie d'un nouveau site. Code que l'on peut directement reprendre de la saisie d'une visite.

**Comment by camillemonchicourt** [comment]
Ce qui existe sur les visites est bien différent.
C'est le fait d'indiquer les observateurs d'une visite.

Mais c'est pas pour dire qui a des droits sur telle ou telle visite.
Même si ensuite la portée des actions des permissions de la visite s'appuie en effet sur les observateurs.

Là ce que tu proposes est de renseigner des personnes au niveau des sites pour les utiliser pour les portées ?

Dans ce cas pourquoi pas, ça reste dans la logique des permissions de GeoNature.

Mais ça fait un champs N-N à ajouter en sur la table des sites ? Comment ça s'appellerait ? Car ce sont pas des observateurs des sites ? Des gestionnaires ? Non plus je pense... 🤔

**Comment by MathRdt** [comment]
Yes, c'est exactement le cas d'usage que l'on souhaite si j'ai bien compris !
Rajouter un utilisateur en tant qu'observateur/gestionnaire pour qu'il ait le droit de voir/éditer ce site par la suite.

Pour le nom je suis ouvert ! Dans notre cas précis, observers nous conviendrait et permettrait de garder la cohérence avec le fonctionnement des visites, vu que ce champ aurait la même fonction chez nous.

Contributeur/Gestionnaire ou autre me paraissent d'autres bons candidats. Mais peut être qu'ils perdraient d'autres futurs développeurs qui voudraient contribuer et verraient un champ avec un nom différent pour les sites et les visites pour une liste d'utilisateurs rattaché à ce site/visite.

**Comment by joelclems** [comment]
Je verrai bien

- dans la table des sites, ajout d'un champ `private` ou `protected` pour indiquer qu'il est protégé
- ajout une table `cor_protected_site_role` pour  lister les personnes qui peuvent accéder au site
- ajout d'une permission (pour les admins)  pour pouvoir accéder aux sites privés sans y être associés

**Comment by camillemonchicourt** [comment]
Ouais je suis pas trop pour mettre en place un truc plus compliqué et spécifique.
On l'a fait dans le module EXPORT et ça complique pas mal la maintenance, mais aussi ça perd les utilisateurs.
Nos permissions étant déjà hyper complexes, si en plus y a des fonctionnements spécifiques et différents du fonctionnement commun dans chaque module, ça devient ingérable selon moi.

Déjà qu'ajouter une notion d'observateurs/contributeurs sur les sites va déjà assez alourdir tout ça.
